pkg/scheduler/plugins/gpushare/pod_handler.go

```diff
@@ -0,0 +1,128 @@
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+func (n *nodeDeviceCache) onPodAdd(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return
+	}
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+	if isGPUPod(podRequest) {
+		n.assume(nil, pod, pod.Spec.NodeName)
+	}
+}
+
+func (n *nodeDeviceCache) onPodUpdate(oldObj, newObj interface{}) {
+	return
+}
+
+func (n *nodeDeviceCache) onPodDelete(obj interface{}) {
+	var pod *corev1.Pod
+	switch t := obj.(type) {
+	case *corev1.Pod:
+		pod = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pod, ok = t.Obj.(*corev1.Pod)
+		if !ok {
+			return
+		}
+	default:
+		return
+	}
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+	if isGPUPod(podRequest) {
+		n.unAssume(nil, pod, pod.Spec.NodeName)
+	}
+}
+
+// add is used to update deviceUsed when there is a new pod created
+func (n *nodeDeviceCache) assume(allocations []*apiext.DeviceAllocation, pod *corev1.Pod, nodeName string) {
+	n.nodeDeviceInfos[nodeName].lock.Lock()
+	defer n.nodeDeviceInfos[nodeName].lock.Unlock()
+
+	if len(allocations) != 0 {
+		for _, allocation := range allocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.Add(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+		}
+	} else {
+		devicesAllocation, err := apiext.GetDevicesAllocation(pod)
+		if err != nil {
+			klog.Errorf("failed to get %v device allocation, err: %v", pod.Name, err)
+			return
+		}
+		gpuAllocations := (*devicesAllocation)[schedulingv1alpha1.GPU]
+		for _, allocation := range gpuAllocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.Add(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+
+		}
+	}
+
+	for minor, usedResources := range *n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU] {
+		(*n.nodeDeviceInfos[nodeName].deviceFree[schedulingv1alpha1.GPU])[minor] = quotav1.Subtract(
+			(*n.nodeDeviceInfos[nodeName].deviceTotal[schedulingv1alpha1.GPU])[minor],
+			usedResources)
+	}
+}
+
+// unAssume is used to update deviceUsed when there is pod deleted
+func (n *nodeDeviceCache) unAssume(allocations []*apiext.DeviceAllocation, pod *corev1.Pod, nodeName string) {
+	n.nodeDeviceInfos[nodeName].lock.Lock()
+	defer n.nodeDeviceInfos[nodeName].lock.Unlock()
+
+	if len(allocations) != 0 {
+		for _, allocation := range allocations {
+			(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+				(*n.nodeDeviceInfos[nodeName].deviceUsed[schedulingv1alpha1.GPU])[int(allocation.Minor)],
+				allocation.Resources)
+		}
+	} else {
+		devicesAllocation, err := apiext.GetDevicesAllocation(pod)
```

Comment:
"devicesAllocation, err := apiext.GetDevicesAllocation(pod)" should be execute outside,  this shouldn't be the cache's logic.  if do so, you can merge if\else